examples/kernel-api.cpp

```diff
@@ -0,0 +1,179 @@
+       RAJA::statement::For<2,RAJA::loop_exec,
+         RAJA::statement::Lambda<2>
+       >
+      >
+    >;
+
+  //Existing kernel API
+  RAJA::kernel<KERNEL_POLICY>(
+    RAJA::make_tuple(RAJA::RangeSegment(0, 3),  // segment tuple...
+                     RAJA::RangeSegment(5, 8),
+                     RAJA::RangeSegment(20, 23)),
+    [=](int i, int , int ) {
+      printf("i = %d \n", i);
+    },
+
+    [=](int, int j, int) {
+      printf("j = %d \n", j);
+    },
+    [=](int, int j, int k) {
+      printf("j, k = %d  %d \n",j, k);
+    });
+
+
+  std::cout<<"----------------------------------------------------\n \n"<<std::endl;
+  std::cout<<"Testing new kernel lambda statement types \n"<<std::endl;
+
+  //Lambda statement format : lambda idx, segment indices, parameter indices (to be tested...)
+  using NEW_POLICY =
+    RAJA::KernelPolicy<
+      RAJA::statement::For<0,RAJA::loop_exec,
+        RAJA::statement::tLambda<0, camp::idx_seq<0>, camp::idx_seq<>>
+      >,
+      RAJA::statement::For<1,RAJA::loop_exec,
+        RAJA::statement::tLambda<1, camp::idx_seq<1>, camp::idx_seq<>>
+      >,
+     RAJA::statement::For<1, RAJA::loop_exec,
+       RAJA::statement::For<2,RAJA::loop_exec,
+         RAJA::statement::tLambda<2, camp::idx_seq<1,2>, camp::idx_seq<>>
+       >
+      >
+    >;
+  // New Kernel API ...
+  RAJA::kernel<NEW_POLICY>(
+    RAJA::make_tuple(RAJA::RangeSegment(0, 3),  // segment tuple...
+                     RAJA::RangeSegment(5, 8),
+                     RAJA::RangeSegment(20, 23)),
+    [=](int i) {
+      printf("i = %d \n", i);
+    },
+
+    [=](int j) {
+      printf("j = %d \n", j);
+    },
+
+    [=](int j, int k) {
+      printf("j, k = %d  %d \n",j, k);
+    });
+
+
+
+  //-----------------------------------------------
+  printf("\n Testing matrix multiplication kernel with new Lambda API ...\n");
+  int N = 10;
+  double *A = memoryManager::allocate<double>(N * N);
+  double *B = memoryManager::allocate<double>(N * N);
+  double *C = memoryManager::allocate<double>(N * N);
+
+  RAJA::View<double, RAJA::Layout<DIM>> Aview(A, N, N);
+  RAJA::View<double, RAJA::Layout<DIM>> Bview(B, N, N);
+  RAJA::View<double, RAJA::Layout<DIM>> Cview(C, N, N);
+
+  for (int row = 0; row < N; ++row) {
+    for (int col = 0; col < N; ++col) {
+      Aview(row, col) = row;
+      Bview(row, col) = col;
+    }
+  }
+
+  using EXEC_POL6 =
+    RAJA::KernelPolicy<
+      RAJA::statement::For<1, RAJA::loop_exec,
+        RAJA::statement::For<0, RAJA::loop_exec,
+          RAJA::statement::tLambda<0,camp::idx_seq<>,camp::idx_seq<0>>,  // dot = 0.0
+          RAJA::statement::For<2, RAJA::loop_exec,
+            RAJA::statement::tLambda<1, camp::idx_seq<0,1,2>, camp::idx_seq<0>> // inner loop: dot += ...
+          >,
+          RAJA::statement::tLambda<2, camp::idx_seq<0,1>, camp::idx_seq<0>>   // set C(row, col) = dot
+        >
+      >
+    >;
+
+  RAJA::kernel_param<EXEC_POL6>(
+    RAJA::make_tuple(RAJA::RangeSegment(0, N),
```

Comment:
We also talked about making RAJA::make_seg_tuple, and RAJA::make_param_tuple methods that generate distinct types. While (slightly) more verbose, I think it would make the code easier to understand. @artv3, @ajkunen thoughts?

Reply:
Why not RAJA::make_tuple_of_segment_looking_thingies(...) and RAJA::make_tuple_of_parameter_looking_thingies(...)?

Reply:
:-B

Reply:
What about having a single tuple consisting of segments and parameters?

Reply:
Personally, I prefer two distinct tuples for code readability and less disruption to current usage. Also, going with one tuple will require more rework of RAJA internals.